File: src/test/java/net/flyingfishflash/loremlist/unit/domain/maintenance/MaintenanceServiceTests.java
package net.flyingfishflash.loremlist.unit.domain.maintenance;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import net.flyingfishflash.loremlist.domain.ServiceResponse;
import net.flyingfishflash.loremlist.domain.association.AssociationRepository;
import net.flyingfishflash.loremlist.domain.exceptions.DomainException;
import net.flyingfishflash.loremlist.domain.lrmitem.LrmItemRepository;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListRepository;
import net.flyingfishflash.loremlist.domain.maintenance.MaintenanceService;
import net.flyingfishflash.loremlist.domain.maintenance.data.DomainPurged;
import org.junit.jupiter.api.Test;

class MaintenanceServiceTests {

  private final AssociationRepository mockAssociationRepository = mock(AssociationRepository.class);
  private final LrmItemRepository mockLrmItemRepository = mock(LrmItemRepository.class);
  private final LrmListRepository mockLrmListRepository = mock(LrmListRepository.class);
  private final MaintenanceService maintenanceService =
      new MaintenanceService(
          mockAssociationRepository, mockLrmItemRepository, mockLrmListRepository);

  @Test
  void domainIsPurged() {
    when(mockAssociationRepository.delete()).thenReturn(997);
    when(mockLrmItemRepository.delete()).thenReturn(998);
    when(mockLrmListRepository.delete()).thenReturn(999);
    ServiceResponse<DomainPurged> serviceResponse = maintenanceService.purge();
    assertThat(serviceResponse.getContent().associationDeletedCount()).isEqualTo(997);
    assertThat(serviceResponse.getContent().itemDeletedCount()).isEqualTo(998);
    assertThat(serviceResponse.getContent().listDeletedCount()).isEqualTo(999);
  }

  @Test
  void domainIsNotPurgedWhenRepositoryThrows() {
    when(mockAssociationRepository.delete()).thenThrow(new RuntimeException("Lorem Ipsum"));
    assertThatThrownBy(maintenanceService::purge)
        .isInstanceOf(DomainException.class)
        .hasMessage("Items, Lists, and Associations could not be purged.");
  }
}

File: src/test/kotlin/net/flyingfishflash/loremlist/unit/domain/maintenance/MaintenanceServiceTests.kt
package net.flyingfishflash.loremlist.unit.domain.maintenance

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.mockk
import io.mockk.unmockkAll
import net.flyingfishflash.loremlist.core.exceptions.ApiException
import net.flyingfishflash.loremlist.domain.association.AssociationRepository
import net.flyingfishflash.loremlist.domain.lrmitem.LrmItemRepository
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListRepository
import net.flyingfishflash.loremlist.domain.maintenance.MaintenanceService

class MaintenanceServiceTests :
  DescribeSpec({

    val mockAssociationRepository = mockk<AssociationRepository>()
    val mockLrmItemRepository = mockk<LrmItemRepository>()
    val mockLrmListRepository = mockk<LrmListRepository>()
    val maintenanceService = MaintenanceService(mockAssociationRepository, mockLrmItemRepository, mockLrmListRepository)

    afterEach { clearAllMocks() }
    afterSpec { unmockkAll() }

    describe("purge()") {
      it("domain is purged") {
        every { mockAssociationRepository.delete() } returns 997
        every { mockLrmItemRepository.delete() } returns 998
        every { mockLrmListRepository.delete() } returns 999
        val serviceResponse = maintenanceService.purge()
        serviceResponse.associationDeletedCount.shouldBe(997)
        serviceResponse.itemDeletedCount.shouldBe(998)
        serviceResponse.listDeletedCount.shouldBe(999)
      }

      it("domain is not purged (repository exception") {
        every { mockAssociationRepository.delete() } throws RuntimeException("Lorem Ipsum")
        val exception = shouldThrow<ApiException> { maintenanceService.purge() }
        exception.message.shouldBe("Items, Lists, and Associations could not be purged.")
      }
    }
  })
